app/WebParsing/UtsgJsonParser.hs

```diff
@@ -0,0 +1,173 @@
+                                                             schedule = fromJust $ HM.lookup "schedule" sectionHash
+                                                             dayLst = lookupVal schedule "meetingDay"
+                                                             startLst = lookupVal schedule "meetingStartTime"
+                                                             endLst = lookupVal schedule "meetingEndTime"
+                                                             timesStr = zipTimeStr dayLst startLst endLst
+                                                             instructors = fromJust $ HM.lookup "instructors" sectionHash
+                                                             instructorStr = zipInstructors (lookupVal instructors "firstName") (lookupVal instructors "lastName")
+                                                             wait = if (T.unpack wait') == "Y" then 0 else -1
+                                                             enrol = 0
+                                                             extra = 0
+                                                         in
+                                                         if (take 3 section) == "LEC"
+                                                         then Left $ Lecture code
+                                                                             session
+                                                                             (T.pack section)
+                                                                             (concat $ zipTimes dayLst startLst endLst)
+                                                                             (read $ T.unpack cap :: Int)
+                                                                             (T.pack $ intercalate ", " instructorStr)
+                                                                             enrol
+                                                                             wait
+                                                                             extra
+                                                                             (T.pack $ intercalate ", " timesStr)
+                                                        else Right $ Tutorial code
+                                                                              (Just session)
+                                                                              (T.pack section)
+                                                                              (concat $ zipTimes dayLst startLst endLst))
+                             (HM.toList meetings)
+
+instance FromJSON Courses where
+  parseJSON = withObject "Courses" $ \o -> do
+    code <- o .: "code"
+    title  <- o .: "courseTitle"
+    description  <- o .: "courseDescription"
+    meetingsObj <- (o .: "meetings" :: Parser (HM.HashMap String Value))
```

Comment:
@Alrighty, I'm looking at this line more closely. My suspicion is that if you have the type annotation be `Parser (HM.HashMap String (Either Lecture Tutorial))`, you'd get an error because `Either Lecture Tutorial` isn't an instance of `FromJSON`.

But if you do `newtype Meeting = Meeting (Either Lecture Tutorial)` then you could make the type `Parser (HM.HashMap String Meeting)`, and then just define a `FromJSON` instance for `Meeting`.

If this is successful, then `HM.values meetingsObj` should have type `[Meeting]`, which bodes well.

Also you might want to look at the Aeson documentation a little more - I believe there's an "optional key" operator to use instead of `(.:)`, which might help for debugging purposes.